cmd/vlog_health_checker/main.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package main
+
+import (
+	"encoding/binary"
+	"flag"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb-binlog/pump/storage"
+	pb "github.com/pingcap/tipb/go-binlog"
+	"github.com/syndtr/goleveldb/leveldb"
+	"go.uber.org/zap"
+)
+
+var tsKeyPrefix = []byte("ts:")
+
+func main() {
+	// Parse args
+	kvDir := flag.String("kv_dir", "", "path of the pump kv directory")
+	repair := flag.Bool("repair", false, "repair the offsets in LevelDB")
+	flag.Parse()
+	path := flag.Arg(0)
+
+	// Open log file
+	log.Info("Target vlog", zap.String("path", path))
+	fid, err := storage.ParseFid(path)
+	if err != nil {
+		log.Fatal("Failed to parse fid", zap.Error(err), zap.String("path", path))
+	}
+	log.Info("Parse fid", zap.Uint32("fid", fid))
+	logFile, err := storage.NewLogFile(fid, path)
+	if err != nil {
+		log.Fatal("Failed to open log", zap.Error(err), zap.String("path", path))
+	}
+	log.Info("Log opened")
+
+	// Open LevelDB
+	metadb, err := storage.OpenMetadataDB(*kvDir, nil)
+	if err != nil {
+		log.Fatal("Failed to open LevelDB", zap.String("path", *kvDir), zap.Error(err))
+	}
+	defer func() {
+		err := metadb.Close()
+		if err != nil {
+			log.Error("Failed to close metadb", zap.Error(err))
+		}
+	}()
+	log.Info("LevelDB opened")
+
+	var corruptionPos int64 = -1
+	checked := make(map[int64]struct{}, 64)
+	logFile.Scan(
+		0,
+		func(vp storage.ValuePointer, record *storage.Record) error {
+			bl, err := record.GetBinlog()
+			if err != nil {
+				log.Fatal("Failed to get binlog", zap.Int64("offset", vp.Offset), zap.Error(err))
+			}
+			// Ignore rollback binlogs
```

Comment:
ignore those only commit ts = 0? may some fake binlog using rollback and commit ts = start ts

Reply:
OK, I'll fix it.